design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+resource) in the Crossplane control cluster. CRUD operations are executed on
+external resources using the API provided by the infrastructure provider.
+However, there are a number of infrastructure providers, with more being created
+everyday, that expose their API's via the Kubernetes API. This document proposes
+a common pattern for stacks that utilize the Kubernetes client as their API.
+Motivation for this design comes from the desire to support [Rook] as a provider
+in Crossplane.
+
+Using the Kubernetes API to create objects in a controller's reconcile loop is
+not a foreign concept in Crossplane. Existing claim reconcilers create a
+Kubernetes object to represent the external resource. However, Kubernetes-native
+providers differ in that the external resource is actually a Kubernetes object
+itself. This concept is also not new in Crossplane, as [complex workloads]
+operate in this manner. `KubernetesApplication` instances are first scheduled to
+a Kubernetes cluster by the scheduler controller, which causes them to pass the
+predicates of the application controller. It uses the template(s) in the
+`KubernetesApplication` instance to create one or more
+`KubernetesApplicationResource` objects, which are then picked up by the
+resource scheduler. It is responsible for actually creating the
+Kubernetes-native (e.g. `Deployment`, `ConfigMap`, etc.) objects in the target
+cluster, then managing them as a managed reconciler does for an external
+resource. In this scenario, the Kubernetes-native objects are the external
+resources.
+
+## Goals
+
+* Create a common pattern that can be used across infrastructure stacks (and
+  potentially the existing complex workload controllers) that utilize the
+  Kubernetes API as the infrastructure provider API
+* Demonstrate the usage of this pattern with Rook
+
+## Differences from `KubernetesApplication`
+
+While the `KubernetesApplication` pattern detailed above provides a starting
+point for this design, a key difference is that `KubernetesApplication` does not
+adhere to the separation of concern structure of other Crossplane resources
+because Kubernetes resources are not modeled in a `claim <-> portable class <->
+non-portable class <-> managed <-> external` workflow. Kubernetes-native
+infrastructure resources, on the other hand, *should* be modeled with the same
+separation of concern that other cloud-provider resources expose.
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to run [CockroachDB] in a Kubernetes
+cluster. Assuming you have installed Rook and the corresponding CockroachDB
```

Comment:
what do you think about scenarios where you have a target cluster running, but you don't actually have the operator running there yet?  I'm wondering if that would be something we can support or have knowledge of.  If we want to deploy a resource to a target cluster, but it's operator isn't yet running, can we automatically install the operator first so that it is then ready to handle any CRDs that we propagate to the target cluster it's running in?

Reply:
We certainly could! It would just need to be built into the `Create()` method of the managed reconciler for the resource I believe.

Reply:
I like this idea, but would suggesting starting by targeting a pre-installed operator then add this in later. I think if we keep this narrowly scoped to begin with there's a good chance this functionality (1-2 Rook based managed resources) could ship super soon.

Reply:
This idea is now captured under future considerations, so I'm going to resolve this thread.